libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp

```diff
@@ -71,11 +137,28 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
 {
     float dt;
 
-    if (_imu._gyro_raw_sample_rates[instance] <= 0) {
-        return;
-    }
+    _update_sensor_rate(_imu._sample_gyro_count[instance], _imu._sample_gyro_start_us[instance],
+                        _imu._gyro_raw_sample_rates[instance]);
 
-    dt = 1.0f / _imu._gyro_raw_sample_rates[instance];
+    /*
+      we have two classes of sensors. FIFO based sensors produce data
+      at a very predictable overall rate, but the data comes in
+      bunches, so we use the provided sample rate for deltaT. Non-FIFO
+      sensors don't bunch up samples, but also tend to vary in actual
+      rate, so we use the provided sample_us to get the deltaT. The
+      difference between the two is whether sample_us is provided.
+     */
```

Comment:
The sample rate for non-FIFO sensors with onboard ADCs (e.g. MPU6000) should not "vary in actual rate", since the sample timing is also generated on the sensor. The fact that the device registers are read at non-uniform intervals should not have any effect on actual sample intervals, even if samples are dropped due to failure to read a register before it is updated. Are there any I2C or SPI IMU sensors in use for which internal ADC timing is dependent on bus transactions?